pkg/controller/cdapmaster/cdapmaster_controller.go

```diff
@@ -20,27 +20,42 @@ import (
 	alpha1 "cdap.io/cdap-operator/pkg/apis/cdap/v1alpha1"
 	"fmt"
 	appsv1 "k8s.io/api/apps/v1"
+	batchv1 "k8s.io/api/batch/v1"
 	corev1 "k8s.io/api/core/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"reflect"
 	gr "sigs.k8s.io/controller-reconciler/pkg/genericreconciler"
 	"sigs.k8s.io/controller-reconciler/pkg/reconciler"
 	"sigs.k8s.io/controller-reconciler/pkg/reconciler/manager/k8s"
 	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"strconv"
 	"strings"
 	"text/template"
+	"time"
 )
 
 const (
 	containerLabel = "cdap.container"
 	// Heap memory related constants
-	javaMinHeapRatio     = float64(0.6)
-	javaReservedNonHeap  = int64(768 * 1024 * 1024)
-	templateDir          = "templates/"
-	deploymentTemplate   = "cdap-deployment.yaml"
-	uiDeploymentTemplate = "cdap-ui-deployment.yaml"
-	statefulSetTemplate  = "cdap-sts.yaml"
-	serviceTemplate      = "cdap-service.yaml"
+	javaMinHeapRatio         = float64(0.6)
+	javaReservedNonHeap      = int64(768 * 1024 * 1024)
+	templateDir              = "templates/"
+	deploymentTemplate       = "cdap-deployment.yaml"
+	uiDeploymentTemplate     = "cdap-ui-deployment.yaml"
+	statefulSetTemplate      = "cdap-sts.yaml"
+	serviceTemplate          = "cdap-service.yaml"
+	upgradeJobTemplate       = "upgrade-job.yaml"
+
+	upgrading                = "upgrading"
+	upgradeFailed            = "upgrade-failed"
+	upgradeSucceededMessage  = "upgrade-succeeded"
```

Comment:
Should this be `upgradeSucceeded`?

Reply:
No, it's just the message that we set upgrading to when it finishes. There are only two cond types, upgrading and upgradeFailed. On success, both are in state false, which is what the CLH should query for.